store/postgres/src/chain_head_listener.rs

```diff
@@ -65,56 +77,65 @@ impl ChainHeadUpdateListener {
     }
 
     fn listen(
+        logger: Logger,
         metrics: Arc<BlockIngestorMetrics>,
         listener: &mut NotificationListener,
-        update_sender: watch::Sender<ChainHeadUpdate>,
+        mut receiver: Receiver<JsonNotification>,
+        watchers: Arc<RwLock<BTreeMap<String, Watcher>>>,
     ) {
         // Process chain head updates in a dedicated task
-        graph::spawn(
-            listener
-                .take_event_stream()
-                .unwrap()
-                .compat()
-                .try_filter_map(move |notification| {
-                    // Create ChainHeadUpdate from JSON
-                    let update: ChainHeadUpdate =
-                        serde_json::from_value(notification.payload.clone()).unwrap_or_else(|_| {
-                            panic!(
-                                "invalid chain head update received from database: {:?}",
-                                notification.payload
-                            )
-                        });
-
-                    // Observe the latest chain_head_number for each network in order to monitor
-                    // block ingestion
-                    metrics.set_chain_head_number(
-                        &update.network_name,
-                        *&update.head_block_number as i64,
-                    );
-                    futures03::future::ok(Some(update))
-                })
-                .try_for_each(move |update| {
-                    futures03::future::ready(update_sender.send(update).map_err(|_| ()))
-                }),
-        );
+        graph::spawn(async move {
+            while let Some(notification) = receiver.recv().await {
+                // Create ChainHeadUpdate from JSON
+                let update: ChainHeadUpdate =
+                    match serde_json::from_value(notification.payload.clone()) {
+                        Ok(update) => update,
+                        Err(e) => {
+                            crit!(
+                                logger,
+                                "invalid chain head update received from database";
```

Comment:
Can you change the message to `Ignoring invalid ...` - that makes it clearer in the log that there shouldn't be any additional action required. Not sure if that really is `crit!` - this is probably only going to happen during updates if the notification format changes. A `warn!` might be enough, especially if successful updates go through later on.